py_src/ipyelk/nx/pipes.py

```diff
@@ -0,0 +1,193 @@
+from typing import Dict, Hashable, Iterator, Optional
+
+import ipywidgets as W
+import networkx as nx
+import traitlets as T
+
+from ipyelk.elements import Edge, HierarchicalElement, Label, Node, Port, Registry
+from ipyelk.elements.serialization import build_edge, build_shape_map, iter_elements
+
+from .. import diagram
+from ..elements import layout_options as opt
+from ..pipes import BrowserTextSizer, ElkJS, Pipe, Pipeline
+
+root_opts = opt.OptionsWidget(
+    options=[
+        opt.HierarchyHandling(),
+    ],
+).value
+label_opts = opt.OptionsWidget(
+    options=[opt.NodeLabelPlacement(horizontal="center", vertical="center")]
+).value
+node_opts = opt.OptionsWidget(
+    options=[
+        opt.NodeSizeConstraints(),
+    ],
+).value
+
+
+class NXSource(W.Widget):
+    graph: nx.MultiDiGraph = T.Instance(nx.MultiDiGraph)
+    hierarchy: nx.DiGraph = T.Instance(nx.DiGraph, allow_none=True)
+
+
+class NetworkxPipe(Pipe):
+    source: NXSource = T.Instance(NXSource, allow_none=True)
+
+    async def run(self):
+        """Run method"""
+        # the things to make the stuff
+        # self.value = elements
+        graph = self.source.graph
+        hierarchy = process_hierarchy(graph, self.source.hierarchy)
+        root: Node = get_root(hierarchy)
+        if not root.layoutOptions:
+            root.layoutOptions = root_opts
+
+        nodes = []
+        for n, d in graph.nodes(data=True):
+            el = from_nx_node(n, d)
+            if not el.layoutOptions:
+                el.layoutOptions = node_opts
+            nodes.append(el)
+            if not el.labels:
+                el.labels.append(Label(text=el.id, layoutOptions=label_opts))
+
+        # add hierarchy nodes
+        for n, d in hierarchy.nodes(data=True):
+            if n not in graph:
+                el = from_nx_node(n, d)
+                nodes.append(el)
+
+        context = Registry()
+        with context:
+            el_map = build_shape_map(*nodes)
+
+            # nest elements based on hierarchical edges
+            for u, v in hierarchy.edges():
+                parent = u if isinstance(u, Node) else el_map[u]
+                child = v if isinstance(v, Node) else el_map[v]
+                parent.children.append(child)
+            for u, v, d in graph.edges(data=True):
```

Comment:
One quick suggestion, I think that if the `graph` is a `nx.MultiDiGraph`, it will fail here because it will have:
`source, target, edge_label, edge_data`.